File: samples/python/18.bot-authentication/dialogs/logout_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import DialogTurnResult, ComponentDialog, DialogContext
from botbuilder.core import BotFrameworkAdapter
from botbuilder.schema import ActivityTypes
from botframework.connector.auth.user_token_client import UserTokenClient


class LogoutDialog(ComponentDialog):
    def __init__(self, dialog_id: str, connection_name: str):
        super(LogoutDialog, self).__init__(dialog_id)

        self.connection_name = connection_name

    async def on_begin_dialog(
        self, inner_dc: DialogContext, options: object
    ) -> DialogTurnResult:
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_begin_dialog(inner_dc, options)

    async def on_continue_dialog(self, inner_dc: DialogContext) -> DialogTurnResult:
        result = await self._interrupt(inner_dc)
        if result:
            return result
        return await super().on_continue_dialog(inner_dc)

    async def _interrupt(self, inner_dc: DialogContext):
        if inner_dc.context.activity.type == ActivityTypes.message:
            text = inner_dc.context.activity.text.lower()
            if text == "logout":
                user_token_client: UserTokenClient = inner_dc.context.turn_state.get(
                    UserTokenClient.__name__, None
                )
                await user_token_client.sign_out_user(
                    inner_dc.context.activity.from_property.id,
                    self.connection_name,
                    inner_dc.context.activity.channel_id,
                )
                await inner_dc.context.send_activity("You have been signed out.")
                return await inner_dc.cancel_all_dialogs()

File: samples/java_springboot/18.bot-authentication/src/main/java/com/microsoft/bot/sample/authentication/LogoutDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.authentication;

import java.util.concurrent.CompletableFuture;

import com.microsoft.bot.builder.BotFrameworkAdapter;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.ComponentDialog;
import com.microsoft.bot.dialogs.DialogContext;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.schema.ActivityTypes;

public class LogoutDialog extends ComponentDialog {

    private final String connectionName;

    public LogoutDialog(String id, String connectionName) {
        super(id);
        this.connectionName = connectionName;
    }


    @Override
    protected CompletableFuture<DialogTurnResult> onBeginDialog(
        DialogContext innerDc, Object options
    ) {
        DialogTurnResult result = interrupt(innerDc).join();
        if (result != null) {
            return CompletableFuture.completedFuture(result);
        }

        return super.onBeginDialog(innerDc, options);
    }

    @Override
    protected CompletableFuture<DialogTurnResult> onContinueDialog(DialogContext innerDc) {
        DialogTurnResult result = interrupt(innerDc).join();
        if (result != null) {
            return CompletableFuture.completedFuture(result);
        }

        return super.onContinueDialog(innerDc);
    }

    private CompletableFuture<DialogTurnResult> interrupt(DialogContext innerDc) {
        if (innerDc.getContext().getActivity().getType().equals(ActivityTypes.MESSAGE)) {
            String text = innerDc.getContext().getActivity().getText().toLowerCase();

            if (text.equals("logout")) {
                // The bot adapter encapsulates the authentication processes.
                BotFrameworkAdapter botAdapter = (BotFrameworkAdapter) innerDc.getContext()
                    .getAdapter();
                botAdapter.signOutUser(innerDc.getContext(), getConnectionName(), null).join();
                innerDc.getContext().sendActivity(MessageFactory.text("You have been signed out."))
                    .join();
                return innerDc.cancelAllDialogs();
            }
        }

        return CompletableFuture.completedFuture(null);
    }

    /**
     * @return the ConnectionName value as a String.
     */
    protected String getConnectionName() {
        return this.connectionName;
    }

}

